File: articles/myers-briggs-type-indicator-history-influence-and-scientific-critique/python/analyze_mbti_typology_vs_traits.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_mbti_typology_vs_traits.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

DIMENSIONS = ["observed_ei", "observed_sn", "observed_tf", "observed_jp"]
LATENT = ["latent_ei", "latent_sn", "latent_tf", "latent_jp"]
RETEST = ["retest_ei", "retest_sn", "retest_tf", "retest_jp"]
REQUIRED = {
    "person_id",
    "assessment_context",
    *LATENT,
    *DIMENSIONS,
    *RETEST,
    "ei_letter",
    "sn_letter",
    "tf_letter",
    "jp_letter",
    "type_code",
    "retest_type_code",
    "type_changed_on_retest",
    "min_absolute_distance_to_boundary",
    "near_boundary",
    "continuous_signal_strength",
    "boundary_risk_score",
    "information_loss_index",
    "collaboration_score",
    "reflective_utility_score",
}

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    pred = X @ beta
    ss_res = float(np.sum((df[y].to_numpy(float) - pred) ** 2))
    ss_tot = float(np.sum((df[y].to_numpy(float) - df[y].mean()) ** 2))
    r2 = 1.0 - ss_res / ss_tot
    return {
        "coefficients": dict(zip(["intercept", *xs], map(float, beta))),
        "r_squared": r2,
    }

df = pd.read_csv(DATA)
missing = REQUIRED - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

type_frequencies = df["type_code"].value_counts().rename_axis("type_code").reset_index(name="n")
type_frequencies["proportion"] = type_frequencies["n"] / len(df)

context_summary = (
    df.groupby("assessment_context")
    .agg(
        n=("person_id", "count"),
        near_boundary_rate=("near_boundary", "mean"),
        type_change_rate=("type_changed_on_retest", "mean"),
        boundary_risk_mean=("boundary_risk_score", "mean"),
        information_loss_mean=("information_loss_index", "mean"),
        collaboration_score_mean=("collaboration_score", "mean"),
        reflective_utility_mean=("reflective_utility_score", "mean"),
    )
    .reset_index()
)

within_type_variation = (
    df.groupby("type_code")[DIMENSIONS + ["collaboration_score", "reflective_utility_score"]]
    .agg(["count", "mean", "std", "min", "max"])
)
within_type_variation.columns = ["_".join(col).strip("_") for col in within_type_variation.columns]
within_type_variation = within_type_variation.reset_index()

boundary_cases = (
    df[df["near_boundary"]]
    .sort_values("min_absolute_distance_to_boundary")
    .loc[
        :,
        [
            "person_id",
            "assessment_context",
            "type_code",
            "retest_type_code",
            "type_changed_on_retest",
            "observed_ei",
            "observed_sn",
            "observed_tf",
            "observed_jp",
            "min_absolute_distance_to_boundary",
            "boundary_risk_score",
            "information_loss_index",
        ],
    ]
)

models = {
    "continuous_collaboration": ols(df, "collaboration_score", DIMENSIONS),
    "continuous_reflective_utility": ols(df, "reflective_utility_score", DIMENSIONS + ["boundary_risk_score"]),
    "category_note": "Categorical type-code modeling should be compared against continuous-dimensional models. This script reports type summaries and continuous OLS models to emphasize information retention.",
    "responsible_use": "Synthetic demonstration only; not for hiring, promotion, clinical assessment, educational placement, relationship matching, legal evaluation, or individual prediction.",
}

type_change_summary = pd.DataFrame(
    {
        "metric": [
            "n",
            "near_boundary_n",
            "near_boundary_rate",
            "type_changed_on_retest_n",
            "type_changed_on_retest_rate",
            "mean_information_loss_index",
            "mean_boundary_risk_score",
        ],
        "value": [
            len(df),
            int(df["near_boundary"].sum()),
            float(df["near_boundary"].mean()),
            int(df["type_changed_on_retest"].sum()),
            float(df["type_changed_on_retest"].mean()),
            float(df["information_loss_index"].mean()),
            float(df["boundary_risk_score"].mean()),
        ],
    }
)

corr_cols = LATENT + DIMENSIONS + RETEST + [
    "min_absolute_distance_to_boundary",
    "continuous_signal_strength",
    "boundary_risk_score",
    "information_loss_index",
    "collaboration_score",
    "reflective_utility_score",
]

type_frequencies.to_csv(OUT / "python_mbti_type_frequencies.csv", index=False)
context_summary.to_csv(OUT / "python_mbti_context_summary.csv", index=False)
within_type_variation.to_csv(OUT / "python_mbti_within_type_variation.csv", index=False)
boundary_cases.to_csv(OUT / "python_mbti_boundary_cases.csv", index=False)
type_change_summary.to_csv(OUT / "python_mbti_type_change_summary.csv", index=False)
df[corr_cols].corr().to_csv(OUT / "python_mbti_dimensional_correlations.csv")
(OUT / "python_mbti_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_mbti_typology_vs_traits.csv", index=False)
print(f"Wrote outputs to {OUT}")

File: articles/myers-briggs-type-indicator-history-influence-and-scientific-critique/r/analyze_mbti_typology_vs_traits.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_mbti_typology_vs_traits.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

type_frequencies <- as.data.frame(table(data$type_code))
names(type_frequencies) <- c("type_code", "n")
type_frequencies$proportion <- type_frequencies$n / nrow(data)

context_summary <- aggregate(
  data[, c("near_boundary", "type_changed_on_retest", "boundary_risk_score", "information_loss_index", "collaboration_score", "reflective_utility_score")],
  by = list(assessment_context = data$assessment_context),
  FUN = mean
)

within_type_variation <- aggregate(
  data[, c("observed_ei", "observed_sn", "observed_tf", "observed_jp", "collaboration_score", "reflective_utility_score")],
  by = list(type_code = data$type_code),
  FUN = function(x) c(mean = mean(x), sd = sd(x), min = min(x), max = max(x))
)

boundary_cases <- data[data$near_boundary == 1, ]
boundary_cases <- boundary_cases[order(boundary_cases$min_absolute_distance_to_boundary), ]

model_continuous <- lm(
  collaboration_score ~ observed_ei + observed_sn + observed_tf + observed_jp,
  data = data
)

model_reflective <- lm(
  reflective_utility_score ~ observed_ei + observed_sn + observed_tf +
    observed_jp + boundary_risk_score,
  data = data
)

model_type <- lm(
  collaboration_score ~ type_code,
  data = data
)

model_comparison <- data.frame(
  model = c("continuous_dimensions", "type_categories"),
  r_squared = c(summary(model_continuous)$r.squared, summary(model_type)$r.squared),
  adjusted_r_squared = c(summary(model_continuous)$adj.r.squared, summary(model_type)$adj.r.squared)
)

type_change_summary <- data.frame(
  metric = c(
    "n",
    "near_boundary_n",
    "near_boundary_rate",
    "type_changed_on_retest_n",
    "type_changed_on_retest_rate",
    "mean_information_loss_index",
    "mean_boundary_risk_score"
  ),
  value = c(
    nrow(data),
    sum(data$near_boundary),
    mean(data$near_boundary),
    sum(data$type_changed_on_retest),
    mean(data$type_changed_on_retest),
    mean(data$information_loss_index),
    mean(data$boundary_risk_score)
  )
)

corr_cols <- c(
  "latent_ei", "latent_sn", "latent_tf", "latent_jp",
  "observed_ei", "observed_sn", "observed_tf", "observed_jp",
  "retest_ei", "retest_sn", "retest_tf", "retest_jp",
  "min_absolute_distance_to_boundary",
  "continuous_signal_strength",
  "boundary_risk_score",
  "information_loss_index",
  "collaboration_score",
  "reflective_utility_score"
)

write.csv(type_frequencies, file.path(outputs, "r_mbti_type_frequencies.csv"), row.names = FALSE)
write.csv(context_summary, file.path(outputs, "r_mbti_context_summary.csv"), row.names = FALSE)
write.csv(within_type_variation, file.path(outputs, "r_mbti_within_type_variation.csv"), row.names = FALSE)
write.csv(head(boundary_cases, 100), file.path(outputs, "r_mbti_boundary_cases_top100.csv"), row.names = FALSE)
write.csv(type_change_summary, file.path(outputs, "r_mbti_type_change_summary.csv"), row.names = FALSE)
write.csv(model_comparison, file.path(outputs, "r_mbti_model_comparison.csv"), row.names = FALSE)
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_mbti_dimensional_correlations.csv"))
capture.output(summary(model_continuous), file = file.path(outputs, "r_continuous_model_summary.txt"))
capture.output(summary(model_type), file = file.path(outputs, "r_type_model_summary.txt"))
capture.output(summary(model_reflective), file = file.path(outputs, "r_reflective_utility_model_summary.txt"))
write.csv(data, file.path(outputs, "r_scored_mbti_typology_vs_traits.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")
